conans/client/downloaders/caching_file_downloader.py

```diff
@@ -43,17 +47,49 @@ def _lock(self, lock_id):
             finally:
                 thread_lock.release()
 
-    def _caching_download(self, url, file_path, md5, sha1, sha256, **kwargs):
-        h = self._get_hash(url, md5, sha1, sha256)
+    def _caching_download(self, url, file_path, md5, sha1, sha256, conanfile: ConanFile, **kwargs):
+        sources_cache = False
+        h = None
+        if conanfile is not None:
+            if sha256:
+                h = sha256
+                sources_cache = True
+            else:
+                ConanOutput()\
```

Comment:
Note for ourselves: we need modes of behavior for the caching, like being able to raise if the thing is not cached, to raise if it cannot be cached because of no sha256, etc. For future work.